src/main/java/com/liorkn/elasticsearch/plugin/VectorScoringPlugin.java

```diff
@@ -13,23 +13,167 @@
  */
 package com.liorkn.elasticsearch.plugin;
 
-import com.liorkn.elasticsearch.service.VectorScoringScriptEngineService;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.nio.DoubleBuffer;
+import java.util.Collection;
+import java.util.Map;
 
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.store.ByteArrayDataInput;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.plugins.ScriptPlugin;
-import org.elasticsearch.script.ScriptEngineService;
-
+import org.elasticsearch.script.ScriptContext;
+import org.elasticsearch.script.ScriptEngine;
+import org.elasticsearch.script.SearchScript;
+import java.util.ArrayList;
 /**
  * This class is instantiated when Elasticsearch loads the plugin for the
  * first time. If you change the name of this plugin, make sure to update
  * src/main/resources/es-plugin.properties file that points to this class.
  */
 public final class VectorScoringPlugin extends Plugin implements ScriptPlugin {
 
-    public final ScriptEngineService getScriptEngineService(Settings settings) {
-        return new VectorScoringScriptEngineService(settings);
+	@Override
+    public ScriptEngine getScriptEngine(Settings settings, Collection<ScriptContext<?>> contexts) {
+        return new VectorScoringPluginEngine();
     }
 
+    /** This {@link ScriptEngine} uses Lucene segment details to implement document scoring based on their similarity with submitted document. */
+    private static class VectorScoringPluginEngine implements ScriptEngine {
+        @Override
+        public String getType() {
+            return "knn";
+        }
+        
+        private static final int DOUBLE_SIZE = 8;
+        
+        @Override
+        public <T> T compile(String scriptName, String scriptSource, ScriptContext<T> context, Map<String, String> params) {
+            
+        	if (context.equals(SearchScript.CONTEXT) == false) {
+                throw new IllegalArgumentException(getType() + " scripts cannot be used for context [" + context.name + "]");
+            }
+            
+        	// we use the script "source" as the script identifier
+            if ("binary_vector_score".equals(scriptSource)) {
```

Comment:
to diminish indentation please turn this if to a fast reject:
if (!"binary_vector_score".equals(scriptSource) {
  throw new IllegalArgumentException("Unknown script name " + scriptSource);
}
// no need for an else section

Reply:
and please put "binary_vector_score" in a final static class member. It will minimize the amount of string allocations